pkg/remoteimageasync/puller.go

```diff
@@ -0,0 +1,71 @@
+package remoteimageasync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+func RunPullerLoop(
+	ctx context.Context,
+	sessionChan chan *PullSession,
+	completedChan chan string,
+) {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done(): // shut down loop
+				// close(completedChan) // the writer is supposed to close channels, but this is messy due to async ops... leaving it open
+				return
+			case ses, ok := <-sessionChan:
+				if !ok { // sessionChan closed, shut down loop
+					return
+				}
+				go func() {
+					klog.V(2).Infof("%s.RunPullerLoop(): asked to pull image %s with timeout %v\n",
+						prefix, ses.image, ses.timeout)
+					ctxCombined, cancelDontCare := context.WithTimeout(ctx, ses.timeout) // combine timeout and shut down signal into one
+					defer cancelDontCare()                                               // IF we exit, this no longer matters. calling to satisfy linter.
+					//NOTE: the logic for "mustPull" is not needed so long as we are not throttling.
+					//      if we DO implement throttling, then additional logic might be required.
+					// mustPull := !cri.hasImage(ses.image)
+					pullStart := time.Now()
+					// if mustPull {
+					// 	klog.V(2).Infof("%s.RunPullerLoop(): image not found, pulling %s\n", prefix, ses.image)
+					// 	cri.pullImage(ses.image, ctx2)
+					// }
+					pullErr := ses.puller.Pull(ctxCombined) //NOTE: relying existing tests or history to veirfy behavior, asyncPull just wraps it
```

Comment:
How is this `Pull` linked to `NodePublishVolume`?

Reply:
note that the above statement
> if that pull activity completes during kubelet's backoff then the image would be present already when the next call to NodePublishVolume is made and no pull is needed
assumes `pullPolicy: IfNotPresent` and would likely be untrue for `pullPolicy: Always` in terms of the logic in `node_server.go`. whether or not another pull occurs would be up to containerd since the `remotepuller.Pull()` method WOULD be invoked.

Reply:
to say this more simply
```
NodePublishVolume(){
  shouldPull := shouldPullLogic()
  puller:=createPuller()
  if !async
    puller.pull(callerContext) // callerContext is 2m from kubelet
  }else{
    session:=asyncPuller.StartPull(puller,pullerTimeout) // pullerTimeout allows pull to run for longer
    asyncPuller.WaitForPull(session,callerContext) // callerContext is 2m from kubelet
  }
}
```
and in the other go routine you pointed to, the async pull activity calls the same `remoteimage.Pull()` method but calls it with the `pullerTimeout`, not the `callerTimeout`. these are run in separate code paths so that kubelet cancelling after 2m does not cancel the activity. the `session` stays operational until `pullerTimeout` or `container-image-csi` pod is shut down or containerd responds with success or error. these outcomes are captured as the result of the session and the session is marked as completed.

any callers blocked on `remoteimageasync.WaitForPull()` then wake up and observe the result of the activity. remember that they can time out and come back and look up the session that's still ongoing.